app/common/ActionSummary.ts

```diff
@@ -125,29 +127,46 @@ export function createEmptyTableDelta(): TableDelta {
 
 /**
  * Distill a summary further, into tabular form, for ease of rendering.
+ *
+ * If context is supplied, in the form of row data for tables, then
+ * it the difference information is overlaid on it.
  */
-export function asTabularDiffs(summary: ActionSummary): TabularDiffs {
+export function asTabularDiffs(summary: ActionSummary,
+                               context?: Record<string, ResultRow[]>): TabularDiffs {
   const allChanges: TabularDiffs = {};
   for (const [tableId, td] of toPairs(summary.tableDeltas)) {
+    const singleTableContext = context?.[tableId];
     const tableChanges: TabularDiff = allChanges[tableId] = {
       header: [],
       cells: [],
     };
-    // swap order to row-dominant for visualization purposes
+    // need order to be row-dominant for visualization purposes.
     const perRow: {[row: number]: {[name: string]: any}} = {};
     const activeCols = new Set<string>();
-    // iterate through the column-dominant representation grist prefers internally
+    const activeColOrder: Array<string> = [];
+    // First add any background context we have been handed.
+    for (const row of singleTableContext || []) {
+      perRow[row.id] = row;
+      for (const col of Object.keys(row)) {
+        if (!activeCols.has(col)) {
+          activeCols.add(col);
+          activeColOrder.push(col);
```

Comment:
Just checking what is the need of this additional array? If only to simulate the `SortedSet` collection than I think it is not needed. Set is implemented as SortedSet (it maintains the order of insertion when you iterate over it).

If not, can you add a comment?